### PR TITLE
AP-9991 # Added `formElementsService.injectFormElementsIntoForms()`

### PR DESCRIPTION
**Pull request by mymattcarroll**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [ ] Have you tested your implementation locally?
- [ ] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

---

> [!NOTE]
> **Medium Risk**
> Changes core form-embedding behavior (cycle handling, cloning, traversal order) used when rendering nested forms; regressions could affect displayed structure or infinite-loop safety, though behavior is heavily tested.
> 
> **Overview**
> Adds **`injectFormElementsIntoForms()`**, an async batch API that resolves embedded `form` / `infoPage` trees via a **`getForm(formId)`** callback, **caches** each id for the whole call, and **mutates** the passed root forms in place while **cloning** loaded form element trees so cached definitions are not walked in place.
> 
> Refactors sync injection around **`resolveEmbeddedFormElement()`** (not found / auth / cycle / continue). **Cycle edges are skipped** without aborting sibling processing, embedded content is taken from **`structuredClone`** of the target form, and embed **stub `elements` are no longer pre-walked** before resolution—fixing incorrect cycle detection when stubs mirror nested graphs.
> 
> Tests cover stub vs injected trees, multi-hop cycles with stubs, batch caching, and reusing **`getForm`** results as later batch roots.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 603ab49c8870c0719e257666db01b4e1d595834b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/formElementsService.ts [hunk 1]
```diff
@@ -326,61 +321,194 @@ function injectFormElements(
...
+  }
+
+  for (const form of forms) {
+    form.elements = await injectFormElementsAsync(
+      form.elements,
+      getFormCached,
+      [form.id],
+      injectAuthenticatedForms,
+    )
+  }
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/formElementsService.ts:460]: Fixed

File: src/formElementsService.ts [hunk 2]
```diff
@@ -310,12 +310,7 @@ function injectFormElements(
   injectAuthenticatedForms: boolean,
 ): FormTypes.FormElement[] {
   return elements.reduce<FormTypes.FormElement[]>((newElements, element) => {
-    if (
-      (element.type === 'page' ||
-        element.type === 'repeatableSet' ||
-        element.type === 'section') &&
-      Array.isArray(element.elements)
-    ) {
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/formElementsService.ts:318]: Fixed

**Comment by divporter** [comment]
is this being used in the API? It's nice to clean up nonetheless but just checking how relevant it is. (No objections if not entirely relevant)

**Comment by mymattcarroll** [comment]
> is this being used in the API? It's nice to clean up nonetheless but just checking how relevant it is. (No objections if not entirely relevant)

The intention is to use this in the API instead of retrieving all forms from the database. The assumptions are
1. Nested forms usage is minimal across all organisations
2. Only a very small number of forms are being nested
3. The same forms are being nested more than once across different forms

**Review by divporter — APPROVED** [review]

**Comment by divporter** [comment]
> > is this being used in the API? It's nice to clean up nonetheless but just checking how relevant it is. (No objections if not entirely relevant)
> 
> The intention is to use this in the API instead of retrieving all forms from the database. The assumptions are
> 
> 1. Nested forms usage is minimal across all organisations
> 2. Only a very small number of forms are being nested
> 3. The same forms are being nested more than once across different forms

I should have read the PR before asking that question 🤦  I'm on board with the strategy.
